### ISSUE TITLE
UnexpectedAdmissionError and pod rejections

### ISSUE THREAD

**Issue by asm582** [openshift/instaslice-operator#140]
When Multiple pods (100 pods) that run for 10 seconds are submitted to the system, we see for few pods getting rejected by the kubelet with an error message:

```
Message:          Pod was rejected: Allocate failed due to no healthy devices present; cannot allocate unhealthy devices nvidia.com/mig-1g.5gb, which is unexpected
```

The above pods are never retried as Kubelet rejects them, see comment here https://github.com/kubernetes/kubernetes/issues/121061#issuecomment-1815996331

There is a small co-relation with the above error and GPU operator pod failing:

```
(base) openstack@netsres62:~/asmalvan/openshift-instaslice/fix_finalizer_ext_res/instaslice-operator$ kubectl get pods -n gpu-operator
NAME                                                         READY   STATUS             RESTARTS        AGE
gpu-feature-discovery-kwf5h                                  0/2     CrashLoopBackOff   57 (12s ago)    9h
gpu-operator-6855b5d55d-kkm7m                                1/1     Running            0               9h
gpu-operator-node-feature-discovery-gc-6dd8c5b44f-dnhh7      1/1     Running            0               9h
gpu-operator-node-feature-discovery-master-f7b95b446-vlwzt   1/1     Running            0               9h
gpu-operator-node-feature-discovery-worker-cw8qz             1/1     Running            0               9h
nvidia-container-toolkit-daemonset-j5tmj                     1/1     Running            0               9h
nvidia-cuda-validator-hj42h                                  0/1     Completed          0               9h
nvidia-dcgm-exporter-b5vhk                                   1/1     Running            0               9h
nvidia-device-plugin-daemonset-fbb4t                         2/2     Running            7 (7m15s ago)   8h
nvidia-operator-validator-qt9x5  
```

the logs of device plugin pod are:

```
I0930 10:01:11.408735     232 main.go:317] Retrieving plugins.
time="2024-09-30T10:01:11Z" level=info msg="Auto-detected mode as 'nvml'"
time="2024-09-30T10:01:11Z" level=info msg="Generating CDI spec for resource: k8s.device-plugin.nvidia.com/gpu"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia0 as /dev/nvidia0"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/dri/card1 as /dev/dri/card1"
time="2024-09-30T10:01:11Z" level=warning msg="Could not locate /dev/dri/controlD65: pattern /dev/dri/controlD65 not found"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/dri/renderD128 as /dev/dri/renderD128"
time="2024-09-30T10:01:11Z" level=warning msg="Failed to evaluate symlink /host/dev/dri/by-path/pci-0000:0e:00.0-card; ignoring"
time="2024-09-30T10:01:11Z" level=warning msg="Failed to evaluate symlink /host/dev/dri/by-path/pci-0000:0e:00.0-render; ignoring"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia1 as /dev/nvidia1"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/dri/card2 as /dev/dri/card2"
time="2024-09-30T10:01:11Z" level=warning msg="Could not locate /dev/dri/controlD66: pattern /dev/dri/controlD66 not found"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/dri/renderD129 as /dev/dri/renderD129"
time="2024-09-30T10:01:11Z" level=warning msg="Failed to evaluate symlink /host/dev/dri/by-path/pci-0000:0f:00.0-card; ignoring"
time="2024-09-30T10:01:11Z" level=warning msg="Failed to evaluate symlink /host/dev/dri/by-path/pci-0000:0f:00.0-render; ignoring"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia0 as /dev/nvidia0"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap66 as /dev/nvidia-caps/nvidia-cap66"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap67 as /dev/nvidia-caps/nvidia-cap67"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia0 as /dev/nvidia0"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap75 as /dev/nvidia-caps/nvidia-cap75"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap76 as /dev/nvidia-caps/nvidia-cap76"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia0 as /dev/nvidia0"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap84 as /dev/nvidia-caps/nvidia-cap84"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap85 as /dev/nvidia-caps/nvidia-cap85"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia0 as /dev/nvidia0"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap93 as /dev/nvidia-caps/nvidia-cap93"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap94 as /dev/nvidia-caps/nvidia-cap94"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia0 as /dev/nvidia0"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap102 as /dev/nvidia-caps/nvidia-cap102"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap103 as /dev/nvidia-caps/nvidia-cap103"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia1 as /dev/nvidia1"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap219 as /dev/nvidia-caps/nvidia-cap219"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap220 as /dev/nvidia-caps/nvidia-cap220"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia1 as /dev/nvidia1"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap228 as /dev/nvidia-caps/nvidia-cap228"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap229 as /dev/nvidia-caps/nvidia-cap229"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia1 as /dev/nvidia1"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap237 as /dev/nvidia-caps/nvidia-cap237"
time="2024-09-30T10:01:11Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap238 as /dev/nvidia-caps/nvidia-cap238"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia1 as /dev/nvidia1"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap246 as /dev/nvidia-caps/nvidia-cap246"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap247 as /dev/nvidia-caps/nvidia-cap247"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia1 as /dev/nvidia1"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap255 as /dev/nvidia-caps/nvidia-cap255"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-caps/nvidia-cap256 as /dev/nvidia-caps/nvidia-cap256"
time="2024-09-30T10:01:12Z" level=info msg="Using driver version 550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-modeset as /dev/nvidia-modeset"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-uvm-tools as /dev/nvidia-uvm-tools"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidia-uvm as /dev/nvidia-uvm"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/dev/nvidiactl as /dev/nvidiactl"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate libnvidia-egl-gbm.so.*.*: pattern libnvidia-egl-gbm.so.*.* not found\n64-bit library libnvidia-egl-gbm.so.*.*: not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate libnvidia-egl-wayland.so.*.*: pattern libnvidia-egl-wayland.so.*.* not found\n64-bit library libnvidia-egl-wayland.so.*.*: not found"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-allocator.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-allocator.so.550.54.14"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate libnvidia-vulkan-producer.so.550.54.14: pattern libnvidia-vulkan-producer.so.550.54.14 not found\n64-bit library libnvidia-vulkan-producer.so.550.54.14: not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate nvidia_drv.so: pattern nvidia_drv.so not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate libglxserver_nvidia.so.550.54.14: pattern libglxserver_nvidia.so.550.54.14 not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate glvnd/egl_vendor.d/10_nvidia.json: pattern glvnd/egl_vendor.d/10_nvidia.json not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate egl/egl_external_platform.d/15_nvidia_gbm.json: pattern egl/egl_external_platform.d/15_nvidia_gbm.json not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate egl/egl_external_platform.d/10_nvidia_wayland.json: pattern egl/egl_external_platform.d/10_nvidia_wayland.json not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate nvidia/nvoptix.bin: pattern nvidia/nvoptix.bin not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate X11/xorg.conf.d/10-nvidia.conf: pattern X11/xorg.conf.d/10-nvidia.conf not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate X11/xorg.conf.d/nvidia-drm-outputclass.conf: pattern X11/xorg.conf.d/nvidia-drm-outputclass.conf not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate vulkan/icd.d/nvidia_icd.json: pattern vulkan/icd.d/nvidia_icd.json not found\npattern vulkan/icd.d/nvidia_icd.json not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate vulkan/icd.d/nvidia_layers.json: pattern vulkan/icd.d/nvidia_layers.json not found\npattern vulkan/icd.d/nvidia_layers.json not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate vulkan/implicit_layer.d/nvidia_layers.json: pattern vulkan/implicit_layer.d/nvidia_layers.json not found\npattern vulkan/implicit_layer.d/nvidia_layers.json not found"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libcuda.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libcuda.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libcudadebugger.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libcudadebugger.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-allocator.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-allocator.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-cfg.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-cfg.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-gpucomp.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-gpucomp.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-ml.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-ml.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-nvvm.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-nvvm.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-pkcs11-openssl3.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-pkcs11-openssl3.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-pkcs11.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-pkcs11.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.550.54.14 as /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.550.54.14"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/run/nvidia-persistenced/socket as /run/nvidia-persistenced/socket"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate /nvidia-fabricmanager/socket: pattern /nvidia-fabricmanager/socket not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate /tmp/nvidia-mps: pattern /tmp/nvidia-mps not found"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/lib/firmware/nvidia/550.54.14/gsp_ga10x.bin as /lib/firmware/nvidia/550.54.14/gsp_ga10x.bin"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/lib/firmware/nvidia/550.54.14/gsp_tu10x.bin as /lib/firmware/nvidia/550.54.14/gsp_tu10x.bin"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/bin/nvidia-smi as /usr/bin/nvidia-smi"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/bin/nvidia-debugdump as /usr/bin/nvidia-debugdump"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/bin/nvidia-persistenced as /usr/bin/nvidia-persistenced"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/bin/nvidia-cuda-mps-control as /usr/bin/nvidia-cuda-mps-control"
time="2024-09-30T10:01:12Z" level=info msg="Selecting /host/usr/bin/nvidia-cuda-mps-server as /usr/bin/nvidia-cuda-mps-server"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate nvidia_drv.so: pattern nvidia_drv.so not found"
time="2024-09-30T10:01:12Z" level=warning msg="Could not locate libglxserver_nvidia.so.550.54.14: pattern libglxserver_nvidia.so.550.54.14 not found"
E0930 10:01:14.144382     232 main.go:149] error starting plugins: error getting plugins: failed to construct NVML resource managers: error building device map: error building device map from config.resources: error building MIG device map: error visiting devices: error visiting device: error visiting device: error visiting MIG device: error visiting MIG device: error getting MIG profile for MIG device at index '(1, 5)': error getting Compute Instance info: ERROR_NO_PERMISSION
```

Due to operator pod failing the capacity and allocatable object are out of sync:

```
Capacity:
  cpu:                            72
  ephemeral-storage:              958798960Ki
  hugepages-1Gi:                  0
  hugepages-2Mi:                  0
  memory:                         1056712468Ki
  nvidia.com/accelerator-memory:  80Gi
  nvidia.com/mig-1g.5gb:          5
  pods:                           110
Allocatable:
  cpu:                            72
  ephemeral-storage:              958798960Ki
  hugepages-1Gi:                  0
  hugepages-2Mi:                  0
  memory:                         1056712468Ki
  nvidia.com/accelerator-memory:  80Gi
  nvidia.com/mig-1g.5gb:          0
  pods:                           110
```

Below is the pod object:

```
(base) openstack@netsres62:~/asmalvan/openshift-instaslice/fix_finalizer_ext_res/instaslice-operator$ kubectl describe pod cuda-vectoradd-73
Name:             cuda-vectoradd-73
Namespace:        default
Priority:         0
Service Account:  default
Node:             kind-control-plane/
Start Time:       Sun, 29 Sep 2024 21:37:39 -0400
Labels:           <none>
Annotations:      <none>
Status:           Failed
Reason:           UnexpectedAdmissionError
Message:          Pod was rejected: Allocate failed due to no healthy devices present; cannot allocate unhealthy devices nvidia.com/mig-1g.5gb, which is unexpected
IP:               
IPs:              <none>
Containers:
  cuda-vectoradd-73:
    Image:      quay.io/tardieu/vectoradd:0.1.0
    Port:       <none>
    Host Port:  <none>
    Command:
      sh
      -c
      sleep 10
    Limits:
      nvidia.com/accelerator-memory:                        5Gi
      nvidia.com/mig-1g.5gb:                                1
      org.instaslice/10d8d7ce-c604-47fc-9a2a-b7ac0f42f199:  1
    Requests:
      nvidia.com/accelerator-memory:                        5Gi
      nvidia.com/mig-1g.5gb:                                1
      org.instaslice/10d8d7ce-c604-47fc-9a2a-b7ac0f42f199:  1
    Environment Variables from:
      10d8d7ce-c604-47fc-9a2a-b7ac0f42f199  ConfigMap  Optional: false
    Environment:                            <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vlg6j (ro)
Volumes:
  kube-api-access-vlg6j:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                    Age    From               Message
  ----     ------                    ----   ----               -------
  Normal   Scheduled                 2m24s  default-scheduler  Successfully assigned default/cuda-vectoradd-73 to kind-control-plane
  Warning  UnexpectedAdmissionError  2m24s  kubelet            Allocate failed due to no healthy devices present; cannot allocate unhealthy devices nvidia.com/mig-1g.5gb, which is unexpected

```

**Comment by asm582** [comment]
Steps to reproduce:

- Use script: https://github.com/openshift/instaslice-operator/blob/main/test/perf/load_with_stats.sh
- Set to 100  https://github.com/openshift/instaslice-operator/blob/5b649d7691eb2898da664a7c9f4b4af726965f41/test/perf/load_with_stats.sh#L11
- change comment to `sleep 10` https://github.com/openshift/instaslice-operator/blob/5b649d7691eb2898da664a7c9f4b4af726965f41/test/perf/load_with_stats.sh#L34
- bash ./perf/load_with_stats.sh

**Comment by asm582** [comment]
This setup is done on a real GPU system and the cluster is spawned using this script: https://github.com/openshift/instaslice-operator/blob/main/deploy/setup.sh

**Comment by asm582** [comment]
Solved by #141
